src/debugger/include/debugger/debugger.h

```diff
@@ -29,47 +30,106 @@ namespace debugger {
 /**
  * Define the debugger class, for breakpoint debugging the CPU and Memory
  */
-class Debugger {
+class DebuggerCore {
 	/**
 	 * Ticks for this object
 	 */
-	unsigned long long ticks;
+	unsigned long long ticks = 0;
 
 	/**
 	 * Gameboy instance
 	 */
 	std::unique_ptr<gameboy::Gameboy> gameboy;
 
 	/**
-	 * Pointer to CPU instance inside gameboy
+	 * A vector container for storing instruction breakpoints
 	 */
-	CPU *cpu;
+	std::vector<Address> breakpoints;
 
 	/**
-	 * Pointer to memory instance inside gameboy
+	 * A bool value to specify whether the debugger has currently encountered a
+	 * breakpoint condition and waiting for the user input. This will reset to
+	 * false once run is called again after a breakpoint.
 	 */
-	Memory *memory;
+	bool is_breaking = false;
 
 	/**
-	 * Pointer to GPU instance inside gameboy
+	 * A vector for storing all the processed breakpoints
 	 */
-	GPU *gpu;
+	std::vector<Address> processed_breakpoints;
 
 	/**
-	 * Pointer to Cartridge instance inside gameboy
+	 * A vector container for storing breakpoints at CPU cycles
 	 */
-	Cartridge *cartridge;
+	std::vector<ClockCycles> cycle_breakpoints;
+
+	/**
+	 * A vector container for storing breakpoints at GameBoy cycles
+	 */
+	std::vector<ClockCycles> tick_breakpoints;
+
+	/**
+	 * A vector container iterator to keep track of breakpoints
+	 */
+	std::vector<Address>::iterator breakpoints_iter = breakpoints.begin();
```

Comment:
Yeah, no. I'll depend on the find operation from now on since the complexity is O(1), it would be an unnecessary space complexity wastage if I kept 3 iterators for the breakpoints.